File: app/src/main/java/com/thewizrd/simpleweather/preferences/OSSCreditsPreference.java
package com.thewizrd.simpleweather.preferences;

import android.content.Context;
import android.text.method.LinkMovementMethod;
import android.util.AttributeSet;
import android.widget.TextView;

import androidx.core.text.HtmlCompat;
import androidx.preference.Preference;
import androidx.preference.PreferenceViewHolder;

import com.thewizrd.shared_resources.AsyncTask;
import com.thewizrd.shared_resources.utils.Colors;
import com.thewizrd.shared_resources.utils.Settings;
import com.thewizrd.shared_resources.utils.UserThemeMode;
import com.thewizrd.simpleweather.R;
import com.thewizrd.simpleweather.helpers.ActivityUtils;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.concurrent.Callable;

public class OSSCreditsPreference extends Preference {

    public OSSCreditsPreference(Context context) {
        super(context);
    }

    public OSSCreditsPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public OSSCreditsPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public OSSCreditsPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    public void onBindViewHolder(PreferenceViewHolder holder) {
        super.onBindViewHolder(holder);

        final Context context = getContext();

        int bg_color = Settings.getUserThemeMode() != UserThemeMode.AMOLED_DARK ?
                ActivityUtils.getColor(context, android.R.attr.colorBackground) : Colors.BLACK;
        holder.itemView.setBackgroundColor(bg_color);

        TextView webView = holder.itemView.findViewById(R.id.textview);

        String creditsText = new AsyncTask<String>().await(new Callable<String>() {
            @Override
            public String call() throws Exception {
                StringBuilder sBuilder = new StringBuilder();
                InputStreamReader sReader = null;
                try {
                    sReader = new InputStreamReader(context.getAssets().open("credits/licenses.html"));

                    int c = 0;
                    while ((c = sReader.read()) != -1) {
                        sBuilder.append((char) c);
                    }
                } catch (IOException ignored) {
                } finally {
                    if (sReader != null)
                        sReader.close();
                }

                return sBuilder.toString();
            }
        });

        webView.setText(HtmlCompat.fromHtml(creditsText.replace("\n", "<br/>"), HtmlCompat.FROM_HTML_MODE_COMPACT));
        webView.setMovementMethod(LinkMovementMethod.getInstance());
    }
}

File: app/src/main/java/com/thewizrd/simpleweather/preferences/OSSCreditsPreference.kt
package com.thewizrd.simpleweather.preferences

import android.content.Context
import android.text.method.LinkMovementMethod
import android.util.AttributeSet
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.text.HtmlCompat
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import com.thewizrd.shared_resources.utils.Colors
import com.thewizrd.shared_resources.utils.ContextUtils.getAttrColor
import com.thewizrd.shared_resources.utils.SettingsManager
import com.thewizrd.shared_resources.utils.UserThemeMode
import com.thewizrd.simpleweather.R
import kotlinx.coroutines.*
import java.io.InputStreamReader

class OSSCreditsPreference : Preference {
    private var loadJob: Job? = null

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)

        val settingsManager = SettingsManager(context.applicationContext)

        val bg_color = if (settingsManager.getUserThemeMode() != UserThemeMode.AMOLED_DARK) {
            context.getAttrColor(android.R.attr.colorBackground)
        } else {
            Colors.BLACK
        }
        holder.itemView.setBackgroundColor(bg_color)

        val webView = holder.itemView.findViewById<TextView>(R.id.textview)
        val progressBar = holder.itemView.findViewById<ProgressBar>(R.id.progressBar)

        loadJob?.cancel()
        loadJob = GlobalScope.launch(Dispatchers.Main.immediate) {
            supervisorScope {
                progressBar.visibility = View.VISIBLE

                val creditsText = withContext(Dispatchers.IO) {
                    val sBuilder = StringBuilder()

                    InputStreamReader(context.assets.open("credits/licenses.html")).use { sReader ->
                        var c: Int
                        while (sReader.read().also { c = it } != -1) {
                            val char = c.toChar()
                            sBuilder.append(if (char == '\n') "<br/>" else char)
                        }
                    }

                    sBuilder.toString()
                }

                ensureActive()

                webView.text = HtmlCompat.fromHtml(creditsText, HtmlCompat.FROM_HTML_MODE_COMPACT)
                webView.movementMethod = LinkMovementMethod.getInstance()

                progressBar.visibility = View.GONE
            }
        }
    }

    override fun onDetached() {
        super.onDetached()
        loadJob?.cancel()
    }

    override fun onPrepareForRemoval() {
        super.onPrepareForRemoval()
        loadJob?.cancel()
    }
}